src/components/tl/ucp/tl_ucp_coll.h

```diff
@@ -207,9 +208,32 @@ ucc_tl_ucp_init_task(ucc_base_coll_args_t *coll_args, ucc_base_team_t *team)
     ucc_tl_ucp_team_t *tl_team = ucc_derived_of(team, ucc_tl_ucp_team_t);
     ucc_tl_ucp_task_t *task    = ucc_tl_ucp_get_task(tl_team);
 
+    if (ucc_unlikely(!task)) {
+        return NULL;
+    }
+
     ucc_coll_task_init(&task->super, coll_args, team);
-    tl_team->seq_num = (tl_team->seq_num + 1) % UCC_TL_UCP_MAX_COLL_TAG;
-    task->tagged.tag           = tl_team->seq_num;
+    if (coll_args->mask & UCC_COLL_ARGS_FIELD_TAG) {
+        task->tagged.tag = coll_args->args.tag;
+    } else {
+        tl_team->seq_num = (tl_team->seq_num + 1) % UCC_TL_UCP_MAX_COLL_TAG;
+        task->tagged.tag = tl_team->seq_num;
+    }
+
+    if (UCC_COLL_ARGS_ACTIVE_SET(&coll_args->args)) {
+        task->tagged.tag = (coll_args->mask & UCC_COLL_ARGS_FIELD_TAG)
+            ? coll_args->args.tag : UCC_TL_UCP_ACTIVE_SET_TAG;
```

Comment:
i don't get it a bit.
In you original code change (Honor user tag - that re-used) you add the "user tag bit" in the send/recv functions by checking mask & COLL_ARGS_FIELD_TAG.

So we have these options:
1. No user tags is provided, then we don't set USER_BIT_FLAG but use pre-defined UCC_TL_UCP_ACTIVE_SET_TAG. This way user can only use the api with "init & post ORDERED" semantics - which is common case. Then the matching of sends and recvs is done based on order (they all have same tag). The active set data is separated since UCC_TL_UCP_ACTIVE_SET_TAG is beyond the range of autogenerated (seq_num) tags.
2. User provides tag. Then we just use it. (the bit will be set in send/recv inlines)

Reply:
How does UCC tell the difference between a collective that is using active sets and custom tags an a non-active set collective?

Doesn't this seem the same to it?
```c
Bcast(tag=0, active_set={0,1})
Bcast(tag=NULL, active_set=NULL)
// Assumes the implicit tags that we use for ordering start at `0`
```